src/cargo/core/resolver/dep_cache.rs

```diff
@@ -164,26 +165,58 @@ impl<'a> RegistryQueryer<'a> {
             }
         }
 
+        // Is this an exact req of the form `=1.0.0` or `=1.0.0+metadata`. We
+        // boost priority of exact matches to such requirements, taking into
+        // account build metadata, which is otherwise not considered relevant in
+        // whether a version matches a requirement.
+        let is_exact_req = match dep.version_req() {
+            OptVersionReq::Any => false,
+            OptVersionReq::Req(req) => {
+                dep.specified_req().is_some()
+                    && req.comparators.len() == 1
+                    && req.comparators[0].op == semver::Op::Exact
+                    && req.comparators[0].patch.is_some()
+            }
+        };
+
+        let prioritize_exact_match = |v: &semver::Version| -> bool {
+            dep.specified_req()
+                .unwrap()
+                .trim()
+                .trim_end_matches(&v.to_string())
```

Comment:
Historically the resolver at least as a whole is definitely performance sensitive, although I don't recall if this exact region is so sensitive to performance. I would expect that we have a lot of `=` exact requirements which typically come from "locked" dependencies and lockfiles (an artifact of how Cargo implemented lock files originally, very dated now...), so this may end up getting executed a lot.

Would you be up for trying to gather some peformance data about the impact of this comparison? I'm mostly worried about the temporarily allocated string here and the various string manipulations, and if they should up at all in comparisons or anything. One of the better ways I've found to benchmark the resolver is to use the `--build-plan` option which avoids actually building anything but still runs the resolver the same way `cargo build` does.

Reply:
Or another way to put this is that this feels somewhat odd doing a bunch of string manipulations here, could the `semver` crate be used instead to extract build metadata and keep that stored somewhere? (or something like that)

Reply:
This code is in a closure that *only* executes if all of the following are true:

- `dep.version_req()` is an `=` requirement.
- `dep.version_req()` gives a specific patch version, i.e. `=x.y.z` or `=x.y.z-pre` or `=x.y.z+meta` or `=x.y.z-pre+meta`, but not `=x.y` or `=x`;
- `ret` contains *multiple* matching candidates that must be sorted. If `ret` contains only 1 possible candidate, then `ret.sort_unstable_by` is never going to run the argument closure.

Because of [how `=` with a specific patch version works](https://github.com/dtolnay/semver/blob/1.0.3/src/eval.rs#L44-L62), the only possible matching candidates (the contents of `ret`) all have `v.major == req.major && v.minor == req.minor && v.patch == req.patch && v.pre == req.pre`, meaning it's only possible for `ret` to contain more than one matching candidate if those candidates all have identical major/minor/patch/pre and differ only in metadata.

So I wouldn't expect a string allocation or string manipulation in this closure to affect resolver performance in general, except when dealing with a large dependency tree consisting overwhelmingly of versions with identical major/minor/patch/pre and differing metadata, which doesn't happen in practice.

For benchmarking do you mostly care about typical dependency trees in which versions don't tend to differ only in metadata (and this closure would therefore execute 0 times), or about worst case performance where every crate has multiple of the same version everywhere?

Reply:
There is a `self.registry_cache` around this method, so it is called at most once per dep per call to resolve. So I would not expect it to be hot.

Reply:
Oh it makes sense that this doesn't really run that often, but Cargo's resolver runs (unintuitively) twice on each build. The first is the "real" one and the latter is a re-resolution to get a smaller set of dependencies (it's weird and it's a bug we've been meaning to fix, well, forever). The implementation with lock files will also synthetically switch many dependencies to `=` dependencies under the hood to ensure the same resolution happens once you already have a lock file (again this is a bug we'd like to fix...)

I think you're right though that the closure probably won't get run since most of the candidate lists will be of length 1. I'm not really worried at all about candidate lists with tons of build metadata and same versions. That being said I'd personally appreciate some smoke testing at least to confirm that this doesn't have a performance impact (as expected).

For performance I'd be interested in big trees, not any particular shape of a tree per se. Something like rust-lang/rust is probably a big enough tree that just doing `cargo build --build-plan` for rustc itself should show a meaningful difference if there is one (and we'd expect no difference here really)